### PR TITLE
add SAI_NEXT_HOP_ATTR_TUNNEL_VNI

### PR DESCRIPTION
**Pull request by lguohan**
This is to per nexthop entry VNI which is used to override the
VNI derived from tunnel mappers

### CODE REVIEW

**Comment by karthik-krishnamurthy** [comment]
@lguohan Would we need VRF to VNI mapper with this change anymore ?

File: inc/sainexthop.h [hunk 1]
```diff
@@ -149,6 +149,26 @@ typedef enum _sai_next_hop_attr_t
...
+     * @validonly SAI_NEXT_HOP_ATTR_TYPE == SAI_NEXT_HOP_TYPE_TUNNEL_ENCAP
+     */
+    SAI_NEXT_HOP_ATTR_TUNNEL_VNI,
+
+    /**
+     * @brief Inner destination MAC address
+     *
+     * @type sai_mac_t
+     * @flags CREATE_AND_SET
+     * @default attrvalue SAI_SWITCH_ATTR_VXLAN_DEFAULT_ROUTER_MAC
```
Comment (kcudnik) [inline inc/sainexthop.h]: SAI_SWITCH_ATTR_VXLAN_DEFAULT_ROUTER_MAC is not defined anywhere

**Comment by karthik-krishnamurthy** [comment]
@lguohan Please let us know the use case for adding tunnel VNI and inner dest mac in next hop . Is it for supporting symmetric IRB in VXLAN ? Also would we need to add the inner tunnel source mac explicitly or can it be same as the outer src mac which i guess would be derived from the RIF

File: inc/sainexthop.h (hunk 1 above)
Comment (lguohan) [reply inc/sainexthop.h]: combined with #751

**Comment by lguohan** [comment]
@karthik-krishnamurthy, the VNI from nexthop is used to override the default one get from the VRF->VNI mapping.

**Comment by lguohan** [comment]
for IRB case, I think by default the src mac can be the overlay rif mac address.

**Comment by tushar-ty** [comment]
Looks like some minor issues in the new version:
a. Data type do not match:
sai_uint32_t vni;
...
attr.id = SAI_TUNNEL_MAP_ENTRY_ATTR_VNI_VALUE;
attr.value.oid = vni;

b. Is this correct:
    attr.id = SAI_TUNNEL_MAP_ENTRY_ATTR_VNI_KEY;
    attr.value.oid = router_id;  <------- vr_id for VNI_KEY ?
    tunnel_map_entry_attrs.push_back(attr);

    attr.id = SAI_TUNNEL_MAP_ENTRY_ATTR_VIRTUAL_ROUTER_VALUE;
    attr.value.oid = vni; <---- vni for VR value as well as data type issue.

File: doc/SAI-Proposal-QinQ-VXLAN.md [hunk 2]
```diff
@@ -348,18 +387,42 @@ Create nexthop for the tunnel interface
...
-  next_hop_attrs[2].id = SAI_NEXT_HOP_ATTR_TUNNEL_ID;
-  next_hop_attrs[2].value.oid = tunnel_id;
-
-  sai_status_t status = sai_next_hop_api->create_next_hop(next_hop_id, router_id, 3, next_hop_attrs);
+  vector<sai_attribute_t> next_hop_attrs;
+
+  sai_attribute_t next_hop_attr;
+
+  next_hop_attr.id = SAI_NEXT_HOP_ATTR_TYPE;
+  next_hop_attr.value.s32 = SAI_NEXT_HOP_TYPE_ENCAP;
```
Comment (prsunny) [inline doc/SAI-Proposal-QinQ-VXLAN.md]: SAI_NEXT_HOP_TYPE_TUNNEL_ENCAP

Comment (lguohan) [reply doc/SAI-Proposal-QinQ-VXLAN.md]: resolved.

File: inc/saiswitch.h [hunk 3]
```diff
@@ -1590,6 +1590,25 @@ typedef enum _sai_switch_attr_t
...
+     * @brief Default VXLAN router MAC
+     *
+     * @type sai_mac_t
+     * @flags CREATE_AND_SET
+     * @default vendor
+     */
+    SAI_SWITCH_ATTR_VXLAN_DEFAULT_ROUTER_MAC,
+
+    /**
+     * @brief Default VXLAN destination port
```
Comment (lguohan) [inline inc/saiswitch.h]: udp port.

File: inc/sainexthop.h [hunk 4]
```diff
@@ -149,6 +149,26 @@ typedef enum _sai_next_hop_attr_t
...
 
+    /**
+     * @brief Next hop entry VNI
+     *
+     * @type sai_uint32_t
+     * @flags CREATE_AND_SET
+     * @default 0
+     * @validonly SAI_NEXT_HOP_ATTR_TYPE == SAI_NEXT_HOP_TYPE_TUNNEL_ENCAP
+     */
+    SAI_NEXT_HOP_ATTR_TUNNEL_VNI,
```
Comment (lguohan) [inline inc/sainexthop.h]: this take precedence over the tunnel mapper.

File: inc/saiswitch.h [hunk 5]
```diff
@@ -1590,6 +1590,25 @@ typedef enum _sai_switch_attr_t
      */
     SAI_SWITCH_ATTR_MIN_BFD_TX,
 
+    /**
+     * @brief Default VXLAN router MAC
```
Comment (lguohan) [inline inc/saiswitch.h]: inner destination mac of the vxlan encap packet.

File: doc/SAI-Proposal-QinQ-VXLAN.md [hunk 6]
```diff
@@ -140,32 +140,142 @@ __Figure 2: QinQ to VxLAN SAI Pipeline__
...
+
+/* add tunnel map entry for C4 */
+status = create_encap_tunnel_map_entry(tunnel_encap_map_id, vrid_3_ingress, 2005);
+status = create_decap_tunnel_map_entry(tunnel_decap_map_id, vrid_3_egress, 2005);
+
+/* create tunnel decap for VM to customer server */
+status = create_tunnel_termination(tunnel_id, "10.10.10.10", &term_table_id);
+
+/* create tunnel nexthop for VM1, VM2 and VM3 */
+status = create_nexthop_tunnel("10.10.10.1", 0, NULL, tunnel_id, &nexthop_id_1);
```
Comment (wendani) [inline doc/SAI-Proposal-QinQ-VXLAN.md]: Need to specify vni 2000 to override tunnel mapper default

**Pull request merged**